arc: codegen or inject destructor error

Appears on nim compiler itself when compiled with `--gc:arc --sinkInference:on`
```nim
type
  TPassContext = object of RootObj
  PPassContext = ref TPassContext

  PCtx = ref object of TPassContext
    a: int

  ModuleGraph = object
    vm: RootRef

proc myOpen(g: sink ModuleGraph): PPassContext = 
  result = PCtx g.vm

var g = ModuleGraph(vm: new(PCtx))

let vm = myOpen(g)
```

Generated C doesn't compile, error:
```
nimcache_z3/@mz3.nim.c:235:3: error: assignment to cast is illegal, lvalue cast are not supported:
        ((tyObject_PCtxcolonObjectType___crveL78K5XzCmJYqrRnx2g*) (g.vm)) = 0;
```

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2567,7 +2567,6 @@ proc upConv(p: BProc, n: PNode, d: var TLoc) =
               [r, checkFor, raiseInstr(p)])
   if n[0].typ.kind != tyObject:
     if n.isLValue:
-      # preserve lvalueness
       putIntoDest(p, d, n,
                 "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
     else:
@@ -2584,8 +2583,6 @@ proc downConv(p: BProc, n: PNode, d: var TLoc) =
   let dest = skipTypes(n.typ, abstractPtrs)
   let src = skipTypes(arg.typ, abstractPtrs)
   discard getTypeDesc(p.module, src)
-  var a: TLoc
-  initLocExpr(p, arg, a)
   let isRef = skipTypes(arg.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}
   if isRef and d.k == locNone and n.typ.skipTypes(abstractInstOwned).kind in {tyRef, tyPtr} and n.isLValue:
     # it can happen that we end up generating '&&x->Sup' here, so we pack
@@ -2593,13 +2590,16 @@ proc downConv(p: BProc, n: PNode, d: var TLoc) =
     # (see bug #837). However sometimes using a temporary is not correct:
     # init(TFigure(my)) # where it is passed to a 'var TFigure'. We test
     # this by ensuring the destination is also a pointer:
-    # preserve lvalueness
+    var a: TLoc
+    initLocExpr(p, arg, a)
     putIntoDest(p, d, n,
               "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
   elif p.module.compileToCpp:
     # C++ implicitly downcasts for us
     expr(p, arg, d)
   else:
+    var a: TLoc
+    initLocExpr(p, arg, a)
     var r = rdLoc(a) & (if isRef: "->Sup" else: ".Sup")
     for i in 2..abs(inheritanceDiff(dest, src)): r.add(".Sup")
     putIntoDest(p, d, n, if isRef: "&" & r else: r, a.storage)

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2605,7 +2605,7 @@ proc downConv(p: BProc, n: PNode, d: var TLoc) =
       # init(TFigure(my)) # where it is passed to a 'var TFigure'. We test
       # this by ensuring the destination is also a pointer:
       if d.k == locNone and skipTypes(n.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}:
-        if n.isLValue:
+        if arg.typ.skipTypes(abstractInstOwned).kind in {tyRef, tyPtr} and n.isLValue:
           # preserve lvalueness
           putIntoDest(p, d, n,
                     "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2578,45 +2578,31 @@ proc upConv(p: BProc, n: PNode, d: var TLoc) =
                         [getTypeDesc(p.module, dest), addrLoc(p.config, a)], a.storage)
 
 proc downConv(p: BProc, n: PNode, d: var TLoc) =
-  if p.module.compileToCpp:
-    discard getTypeDesc(p.module, skipTypes(n[0].typ, abstractPtrs))
-    expr(p, n[0], d)     # downcast does C++ for us
+  var arg = n[0]
+  while arg.kind == nkObjDownConv: arg = arg[0]
+
+  let dest = skipTypes(n.typ, abstractPtrs)
+  let src = skipTypes(arg.typ, abstractPtrs)
+  discard getTypeDesc(p.module, src)
+  var a: TLoc
+  initLocExpr(p, arg, a)
+  let isRef = skipTypes(arg.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}
+  if isRef and d.k == locNone and n.typ.skipTypes(abstractInstOwned).kind in {tyRef, tyPtr} and n.isLValue:
+    # it can happen that we end up generating '&&x->Sup' here, so we pack
+    # the '&x->Sup' into a temporary and then those address is taken
+    # (see bug #837). However sometimes using a temporary is not correct:
+    # init(TFigure(my)) # where it is passed to a 'var TFigure'. We test
+    # this by ensuring the destination is also a pointer:
+    # preserve lvalueness
+    putIntoDest(p, d, n,
+              "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
+  elif p.module.compileToCpp:
+    # C++ implicitly downcasts for us
+    expr(p, arg, d)
   else:
-    var dest = skipTypes(n.typ, abstractPtrs)
-
-    var arg = n[0]
-    while arg.kind == nkObjDownConv: arg = arg[0]
-
-    var src = skipTypes(arg.typ, abstractPtrs)
-    discard getTypeDesc(p.module, src)
-    var a: TLoc
-    initLocExpr(p, arg, a)
-    var r = rdLoc(a)
-    let isRef = skipTypes(arg.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}
-    if isRef:
-      r.add("->Sup")
-    else:
-      r.add(".Sup")
+    var r = rdLoc(a) & (if isRef: "->Sup" else: ".Sup")
     for i in 2..abs(inheritanceDiff(dest, src)): r.add(".Sup")
-    if isRef:
-      # it can happen that we end up generating '&&x->Sup' here, so we pack
-      # the '&x->Sup' into a temporary and then those address is taken
-      # (see bug #837). However sometimes using a temporary is not correct:
-      # init(TFigure(my)) # where it is passed to a 'var TFigure'. We test
-      # this by ensuring the destination is also a pointer:
-      if d.k == locNone and skipTypes(n.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}:
-        if arg.typ.skipTypes(abstractInstOwned).kind in {tyRef, tyPtr} and n.isLValue:
-          # preserve lvalueness
-          putIntoDest(p, d, n,
-                    "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
-        else:
-          getTemp(p, n.typ, d)
-          linefmt(p, cpsStmts, "$1 = &$2;$n", [rdLoc(d), r])
-      else:
-        r = "&" & r
-        putIntoDest(p, d, n, r, a.storage)
-    else:
-      putIntoDest(p, d, n, r, a.storage)
+    putIntoDest(p, d, n, if isRef: "&" & r else: r, a.storage)
 
 proc exprComplexConst(p: BProc, n: PNode, d: var TLoc) =
   let t = n.typ

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -2567,6 +2567,7 @@ proc upConv(p: BProc, n: PNode, d: var TLoc) =
               [r, checkFor, raiseInstr(p)])
   if n[0].typ.kind != tyObject:
     if n.isLValue:
+      # preserve lvalueness
       putIntoDest(p, d, n,
                 "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
     else:
@@ -2604,8 +2605,13 @@ proc downConv(p: BProc, n: PNode, d: var TLoc) =
       # init(TFigure(my)) # where it is passed to a 'var TFigure'. We test
       # this by ensuring the destination is also a pointer:
       if d.k == locNone and skipTypes(n.typ, abstractInstOwned).kind in {tyRef, tyPtr, tyVar, tyLent}:
-        getTemp(p, n.typ, d)
-        linefmt(p, cpsStmts, "$1 = &$2;$n", [rdLoc(d), r])
+        if n.isLValue:
+          # preserve lvalueness
+          putIntoDest(p, d, n,
+                    "(*(($1*) (&($2))))" % [getTypeDesc(p.module, n.typ), rdLoc(a)], a.storage)
+        else:
+          getTemp(p, n.typ, d)
+          linefmt(p, cpsStmts, "$1 = &$2;$n", [rdLoc(d), r])
       else:
         r = "&" & r
         putIntoDest(p, d, n, r, a.storage)

--- a/compiler/renderer.nim
+++ b/compiler/renderer.nim
@@ -1143,9 +1143,9 @@ proc gsub(g: var TSrcGen, n: PNode, c: TContext) =
     gcomma(g, n)
     put(g, tkParRi, ")")
   of nkObjDownConv, nkObjUpConv:
+    let typ = if (n.typ != nil) and (n.typ.sym != nil): n.typ.sym.name.s else: ""
+    put(g, tkParLe, typ & "(")
     if n.len >= 1: gsub(g, n[0])
-    put(g, tkParLe, "(")
-    gcomma(g, n, 1)
     put(g, tkParRi, ")")
   of nkClosedSymChoice, nkOpenSymChoice:
     if renderIds in g.flags:

--- a/tests/ccgbugs/tlvalueconv.nim
+++ b/tests/ccgbugs/tlvalueconv.nim
@@ -1,3 +1,9 @@
+discard """
+  matrix: "--gc:refc; --gc:arc"
+"""
+
+# bug #14160
+
 type
   TPassContext = object of RootObj
   PPassContext = ref TPassContext
@@ -13,3 +19,13 @@ proc main() =
   PCtx(g.vm) = nil #This generates invalid C code
 
 main()
+
+# bug #14325
+
+proc main2() =
+  var g = ModuleGraph(vm: new(Pctx))
+  PPassContext(PCtx(g.vm)) = nil #This compiles, but crashes at runtime with gc:arc
+  doAssert g.vm == nil
+
+main2()
+

--- a/tests/ccgbugs/tlvalueconv.nim
+++ b/tests/ccgbugs/tlvalueconv.nim
@@ -17,6 +17,7 @@ type
 proc main() =
   var g = ModuleGraph(vm: new(Pctx))
   PCtx(g.vm) = nil #This generates invalid C code
+  doAssert g.vm == nil
 
 main()
 

